### ISSUE TITLE
BUG: IntervalArray.__setitem__ creates copies incorrectly

### ISSUE THREAD

**Issue by jbrockmendel** [pandas-dev/pandas#27147]
```
from pandas.core.arrays import IntervalArray
import numpy as np
arr = np.random.randn(10)
a = IntervalArray.from_arrays(arr, arr+1)
b = a[:5]

b[0] = b[1]
assert a[0] == a[1]  # <-fails
```

**Comment by jbrockmendel** [comment]
I originally thought #32782 would fix this, but it doesn't get all the way across the finish line.

**Comment by TomAugspurger** [comment]
This from the explicit copies at https://github.com/pandas-dev/pandas/blob/72aed3e3052ef92005408a40c34425cd1f09fadd/pandas/core/arrays/interval.py#L556-L564

To check that, this diff gets the original example working.

```diff
diff --git a/pandas/core/arrays/interval.py b/pandas/core/arrays/interval.py
index c861d25afd..344538d1a2 100644
--- a/pandas/core/arrays/interval.py
+++ b/pandas/core/arrays/interval.py
@@ -555,11 +555,13 @@ class IntervalArray(IntervalMixin, ExtensionArray):
 
         # Need to ensure that left and right are updated atomically, so we're
         # forced to copy, update the copy, and swap in the new values.
-        left = self.left.copy(deep=True)
+        # left = self.left.copy(deep=True)
+        left = self.left
         left._values[key] = value_left
         self._left = left
 
-        right = self.right.copy(deep=True)
+        # right = self.right.copy(deep=True)
+        right = self.right
         right._values[key] = value_right
         self._right = right
 ```

However, as the comment indicates I don't think we can just do that. Perhaps the flow should be

```
left2 = self.left.copy()
left2._values[key] = value_left

right2 = self.right.copy()
right2._values[key] = value_right

# We know the setitem works
left = self.left
left._values[key] = value_left
right = self.right
right._values[key] = value_right
```

(the fact that we're manipulating `Index._values` makes me a bit worried, but that's true regardless)

**Comment by jbrockmendel** [comment]
> Perhaps the flow should be

That looks right to me.  Ideally we could do this without making copies, but this is a strict improvement over the status quo.

**Comment by jbrockmendel** [comment]
closed by #36310
